cpp/include/raft/spatial/knn/detail/ann_quantized.cuh

```diff
@@ -0,0 +1,285 @@
+  config.interleavedLayout               = params.n_bits != 8;
+  faiss::MetricType faiss_metric         = build_faiss_metric(params.metric);
+  faiss::gpu::GpuIndexIVFPQ* faiss_index = new faiss::gpu::GpuIndexIVFPQ(
+    index->gpu_res, D, params.n_lists, params.n_subquantizers, params.n_bits, faiss_metric, config);
+  index->index = faiss_index;
+}
+
+template <typename IntType = int>
+void approx_knn_ivfsq_build_index(knnIndex* index,
+                                  const ivf_sq_index_params& params,
+                                  IntType n,
+                                  IntType D)
+{
+  faiss::gpu::GpuIndexIVFScalarQuantizerConfig config;
+  config.device                                       = index->device;
+  faiss::MetricType faiss_metric                      = build_faiss_metric(params.metric);
+  faiss::ScalarQuantizer::QuantizerType faiss_qtype   = build_faiss_qtype(params.qtype);
+  faiss::gpu::GpuIndexIVFScalarQuantizer* faiss_index = new faiss::gpu::GpuIndexIVFScalarQuantizer(
+    index->gpu_res, D, params.n_lists, faiss_qtype, faiss_metric, params.encode_residual);
+  index->index = faiss_index;
+}
+
+template <typename T = float, typename IntType = int>
+void approx_knn_build_index(const handle_t& handle,
+                            knnIndex* index,
+                            const knn_index_params& params,
+                            T* index_array,
+                            IntType n,
+                            IntType D)
+{
+  auto stream      = handle.get_stream();
+  auto metric      = params.metric;
+  index->index     = nullptr;
+  index->metric    = metric;
+  index->metricArg = params.metric_arg;
+  int device;
+  RAFT_CUDA_TRY(cudaGetDevice(&device));
+  index->device    = device;
+  auto ivf_ft_pams = dynamic_cast<const ivf_flat::index_params*>(&params);
+  auto ivf_pq_pams = dynamic_cast<const ivf_pq_index_params*>(&params);
+  auto ivf_sq_pams = dynamic_cast<const ivf_sq_index_params*>(&params);
+
+  // perform preprocessing
+  // k set to 0 (unused during preprocessing / revertion)
+  if constexpr (std::is_same<T, uint8_t>{} || std::is_same<T, int8_t>{}) {
+    if (ivf_ft_pams) {
+      approx_knn_cuivfl_ivfflat_build_index(handle, index, *ivf_ft_pams, index_array, n, D);
+    } else {
+      RAFT_FAIL("IVF Flat algorithm required to fit int8 data");
+    }
+  } else if constexpr (std::is_same<T, float>{}) {
+    std::unique_ptr<MetricProcessor<float>> query_metric_processor =
+      create_processor<float>(metric, n, D, 0, false, stream);
+
+    if (ivf_ft_pams) {
+      // cuivfl only supports L2/Inner product for now.
+      if (metric == raft::distance::DistanceType::L2SqrtExpanded ||
+          metric == raft::distance::DistanceType::L2SqrtUnexpanded ||
+          metric == raft::distance::DistanceType::L2Unexpanded ||
+          metric == raft::distance::DistanceType::L2Expanded ||
+          metric == raft::distance::DistanceType::InnerProduct) {
```

Comment:
Currently the [python layer accepts](https://github.com/rapidsai/cuml/blob/6f51c1fd86af337ca1cddb02ce8e03e25bb9df2d/python/cuml/neighbors/__init__.py#L42) cosine similarity and correlation metrics as well. Please add it to the todo list to support them.